### ISSUE TITLE
Hard Permission System

### ISSUE THREAD

**Issue by LexiBasilisk** [Yellow-Dog-Man/Resonite-Issues#1103]
### Is your feature request related to a problem? Please describe.

Resonite's existing permission system allows for some amount of control over the user experience, but has some pretty significant limitations.
- Most of the existing permissions are binary- either enabling, or disabling functionality outright, e.g. object, and avatar spawning
- World permissions only exist on a per-world basis, and there is no way to easily ensure a consistent experience across worlds the user hosts (particularly for ones they are not the original creator of)
- User-made tooling spawned into sessions can disrupt/break consent of users, and be used in harassing ways, e.g. anchoring the user, applying force to the user, etc.

### Describe the solution you'd like

A hard permission system which among other things would need:
- Low level write/modify validations of the data model, locking down parts of it from being modified by users (even in case of malicious client which bypasses the restrictions on user's end)
- Throttling system - limiting how many objects, components and assets can be spawned/modified per unit of time, preventing people from spamming
- Type filtering - will limit which datatypes can be introduced into the model for loaded/spawned objects. This will allow filtering out certain components/nodes from avatars, objects, so they can still be brought in, but with restricted functionality. It'll also allow users to spawn certain items, but not others.

This functionality would ideally come with UI so users, and world hosts can set, and establish preferences/defaults for the sorts of experiences they wish to have grouped into various categories, e.g. preventing force from being applied, being anchored, etc.

These permissions would need a few layers of control, e.g. a world creator with a world that has built-in anchors would need to be separated, and distinguished from objects that are added into the world during a session. Users that are entering hosted sessions that do not match their desired comfort level would ideally be warned prior to joining the session that the world contains types of content they have opted out from.

Preferences should be able to be set under certain criteria as well, e.g. having more permissive permissions for contacts, compared to non-contacts. Whenever we further rework the contacts UI, this could also integrate deeper, e.g. assigning permissions to a group of users, or on a per-user basis rather than just broad 'roles' that are controlled solely by the hosted world.

### Describe alternatives you've considered

Currently there is no alternative- users that want to ensure a controlled and restricted experience need to tune the existing permission system to meet their needs as best they are able on every single session/world they wish to host.

### Additional Context

There are several issues, e.g. #28, #337, #815, #1100, and possibly others which all broadly would be resolved, or significantly improved by the addition of this feature.

This feature would also help ensure additional user control in choosing the types of experiences they are comfortable with across all of the sessions they host/visit.

**Comment by 989onan** [comment]
What would be nice if the client can block things happening to them, rather than a world blocking them for everyone
I find an issue can arise where a group of users wants to shoot each other or have fun in a hidden location not bothering anyone, but the world says no.

Those users would move somewhere else, or further schism the community into more specialized worlds
What would be nice is if the data model can prevent things from working that users can opt out of themselves, without ruining the experience for everyone (as you mentioned, "preventing force from being applied, being anchored, etc.")

I have made mods to prevent knock back, and in game systems already exist to prevent knock back tools and anchoring from working. It would be preferable to allow users to choose what happens to them, rather than the host choosing for them in such cases. 

But case in point, the host should be able to decide what can be done to the world in a more granular way as you mentioned, but I think it should still allow users to decide to filter the content that only concerns themselves without host intervention

**Comment by Frooxius** [comment]
The goal of this system would be to provide ability for both. More control on what happens to you and your avatar and more control to the host on what happens in their world.

This will allow host to allow that kind of behavior, but individual users still preventing it if they want to.

However we also want to allow the host to block certain things completely if they wish to do so - that might be necessary for certain types of sessions (e.g. game worlds or chill social sessions).

**Comment by JackTheFoxOtter** [comment]
Will the world also be able to "enforce" something to be enabled? For example, a game world that requires knockback wouldn't work for users that disabled knockback affecting them.

**Comment by 989onan** [comment]
> Will the world also be able to "enforce" something to be enabled? For example, a game world that requires knockback wouldn't work for users that disabled knockback affecting them.

would be good to notify users as such, and that certain features may not work and make sure they know that it can very easily break stuff in said world

**Comment by gentlecolts** [comment]
I think one of my personal takeaways from the discussion in #337 was that it might be good to look to web browsers or systems such as iOS/Android for inspiration on handling granular permissions.

For example, at the world level, a world may want to say "This world needs xyz permissions", which would require users accepting (either through global settings or on a per "app" basis), before joining/creating sessions.

It may also be necessary for there to be some sense of "loaded or spawned by the world" and separating that from content spawned by users.  One elegant solution here may be through recursive permissions on slots, where existing workspace managers could have sandboxing restrictions set on their slots, and all user roots under them would be subject to those restrictions.  Content spawned within, therefore, could be made to not be able to even see further out than their own user workspace.  This brings some challenges as colliders or raycasting potentially allowing content to break free from that sandbox, but could still be a very worthwhile avenue for exploration.

But there's also several different perspectives to think about here too:
 - a world, for example, could be treated many different ways, but its permissions are likely to be thought of as "things i need to be able to function properly", some of which may be required, some of which may be optional
 - a session, of which the session's host may have preferences on what should/shouldn't be allowed in the session they're running
 - the user themselves, who may dislike exposure to certain kinds of content, or have a need to restrict some types of content for accessibility reasons, all of which may not necessarily be reasons to prevent them from entering a world/session if it's acceptable for that kind of content to simply be disabled client-side

Another item for consideration is that permissions and discussion around it may have overlap with content performance and content age ratings, and it may be a good idea to draw a clear line about where those subjects overlap

**Comment by EmergencyTemporalShift** [comment]
Something that comes to mind is that, for example if you had a session where knockback was required to participate in the game, spectators would not need to allow knockback on themselves. (Also if you had some sort of unequal team game, where knockback is allowed one way but not the other (think 1 v 3 Mario party games) or a no knockback mode on a game. Would there still be a warning before entering the game world or just before joining a game with those things enabled. Would it be more difficult to enable or disable pre join warnings via flux or components automatically?

**Comment by gentlecolts** [comment]
> Would there still be a warning before entering the game world

Better to ask ahead of time if the user "might" be exposed to a feature by being present in a session than to disrupt experience mid-session to make a request just-in-time, my two cents.

**Comment by charlie-sans** [comment]
from what i'm reading in this issue, the hard perm system is simalar to something like linux and it's kernel rings where the low level parts are rings 1 and 0 and everything else is userspace/ring3?

**Comment by tgiddings** [comment]
This thread mentions multiple levels of permission, as well as ideas like world requirements. I would like to make a specific suggestion of how all of these connect and interact:

- both the user and the session specify permissions. If an action is not allowed by session permissions, or is not allowed by user permissions of the targeted user (if one exists), then the action fails.
- The session and user permissions can be specified depending on who initiates the action, such as the world, users in general, a specific group of users (like friends), etc. If session or user permissions allow an action in any of the groups the initiator belongs to, it allows the action for the initiator.
- if an action is initiated by a spawned object, then for the purposes of permissions, the action is initiated by whoever spawned the object -- whether a user or the world.
- if an action is initiated by an avatar, then for the purposes of permissions, the action is initiated by whichever user is using that avatar.
- if an action targets an object which is part of a user's avatar, then for the purposes of permissions, the action targets the user.
- if an action was initiated by a user, the targeted user has blocked the initiating user, the action fails, regardless of other factors.
- Worlds:
  - specifies the default session permissions of the world
  - specifies the required permissions for normal functioning
- Hosts:
  - can edit the session permissions. the UI should make the host aware if they restrict a permission required by the world, or if they remove a restriction originally set by the world. In my opinion, the game should give deference to the host over the world author in this situation by allowing the permission change, albeit with an appropriate warning. 
  - specify their default session permissions for sessions they create
- Users:
  - can edit their own user permissions.
  - a restriction in user permissions cannot be overridden by other users (including the host) or the world.
  - When joining a world, if their restrictions conflict with the world's requirements, the user is warned that the world may not function correctly. In my opinion, the game should not override the user's restrictions if they choose to enter the world, as I consider it more important to respect a user's boundaries and consent than to guarantee the correct functioning of a world.
- sessions:
  - when created, if a permission is restricted by either the host's or the world's default session permissions, they are restricted in the session's session permission
  - when creating, if a permission is restricted by the host's default session permissions, but required by the world, the host should be warned that the world may not function correctly.

Unanswered questions:
- How should permissions work if they target an object spawned by a user? Presumably, a user should be able to separately say "it's okay to do this to my stuff" and "it's okay to do this to me"
- how should tools spawned by the world be treated? Some tools may simply be additional ways to interact in general, and should be subject to permissions restricting the user. Other tools may be for performing game actions, and should be subject to the permissions restricting the world

Potential enhancements:
- A world may take action on a user on behalf of another user, as part of gameplay. It may be beneficial to give a world the ability to query whether a user would be permitted to take a particular action on the targeted user, or to query if a user is blocked by the targeted user. Access to this information should be gated behind a permission on the targeted user. The game should not try to enforce a specific use of this information, since the conceptual difference between "X user did Y to me" and "the game did Y to me because X did something in the game" is very world-dependent and can vary between different situations in the world. The same and similar data can be used for world logic, such as only allowing a player to spectate a game if they do not grant the required permissions to participate.
- when joining a world that requires permissions which they restrict, a user could be given the option to grant those permissions for just that session or just that world
- when creating a world that requires permissions which their default session permissions restrict, a host could be given the option to grant those permissions for just that session or just that world
- a world could have suggested permissions, which give an individualized warning on create/join . It would likely be beneficial to allow a user to blanket suppress these warnings, since they are merely suggestions

**Comment by JackTheFoxOtter** [comment]
> a restriction in user permissions cannot be overridden by other users (including the host) or the world.
When joining a world, if their restrictions conflict with the world's requirements, the user is warned that the world may not function correctly. In my opinion, the game should not override the user's restrictions if they choose to enter the world, as I consider it more important to respect a user's boundaries and consent than to guarantee the correct functioning of a world.

Depending on what permissions we're talking about, this can lead to content breakage and negatively impact other people in the session as well as the affected user (for example in game worlds / curated experiences). Sessions / worlds need to be able to override this, but in that case users should be informed and ask if they really want to join.

**Comment by GrandtheUK** [comment]
Not exactly hard permissions related but would likely be a layer on top of such a system. but for #57 it could be useful to limit access to certain fields on certain components (i.e. meshrenderer, skinnedmeshrender) to no-access/read-only/read-write to try to prevent asset theft via protoflux grabbing of assets on components. obviously this wouldn't necessarily stop access via inspector when you have builder since that would break things that shouldn't be broken.

Would this be compatible with the hard permission system and would it be something worth looking into?

(if i should make another issue and refer back to this one please let me know)

**Comment by EnergyBun** [comment]
If a user needs to disable flashing effects because of epilepsy, that should be covered in accessibility options, and aside from accessibilities, I hold the philosophy that Resonite should be used for making games, and games, after initial install, should not ask the user for permissions.
